home_energy_management/device_simulators/electric_vehicle.py

```diff
@@ -0,0 +1,179 @@
+            "min_charge_level": self.min_charge_level,
+            "curr_charge_level": self.curr_capacity / self.max_capacity * 100,
+            "nominal_power": self.max_power,
+            "efficiency": self.efficiency,
+            "is_available": self.is_available,
+            "driving_power": self.get_driving_power(now),
+            "next_update_time": self.get_next_update_time(now),
+        }
+
+    def set_params(self, params: dict[str, Any]) -> None:
+        self.update_capacity(self.get_time())
+        self.max_charge_rate = params["InWRte"] / 100.
+        self.max_discharge_rate = params["OutWRte"] / 100.
+        self.operation_mode = params["StorCtl_Mod"]
+        self.notify()
+
+    def adjust_current_to_state(self, info: InfoForDevice) -> None:
+        self.current[0] = 0
+
+        if self.get_driving_power(info.now) > 0:
+            self.is_available = False
+            self.operation_mode = 0
+            self.voltage[0] = 0
+            return
+
+        self.is_available = True
+        self.operation_mode = 1  # TODO - only while no decision from algorithm
+
+        self.voltage = info.voltage
+        if info.voltage[0] == 0:
+            return
+
+        if self.operation_mode == 2:
+            if self.curr_capacity <= self.min_charge_level / 100 * self.max_capacity:
+                self.current[0] = 0
+            else:
+                self.current[0] = self.max_discharge_rate * self.max_power * 1000 / info.voltage[0].real
+        if self.operation_mode == 1:
+            if self.curr_capacity == self.max_capacity:
+                self.current[0] = 0
+            else:
+                power_reduction = min(
+                    self.max_charge_rate,
+                    (self.max_capacity - self.curr_capacity) / (
+                                self.max_capacity * (1 - self.charging_switch_level / 100))
+                )
+                self.current[0] = power_reduction * self.max_power * 1000 / info.voltage[0].real
+
+    def update(self, info: InfoForDevice) -> DeviceResponse:
+        self.adjust_current_to_state(info)
+        return DeviceResponse(self.current, self.update_capacity(info.now))
+
+
+class AbstractEVDriving(ABC):
+    @abstractmethod
+    def get_driving_power(self, now: int) -> float:
+        pass
+
+    @abstractmethod
+    def get_next_update_time(self, now: int) -> int:
+        pass
+
+
+class ScheduledEVDriving(AbstractEVDriving, ScheduledDevice):
+    def __init__(
+            self,
+            config: list[tuple[float, Any]],
+            loop: int = 0):
+        super().__init__(config, loop)
+
+    def get_driving_power(self, now: int) -> float:
+        driving_power, _ = self.get_state(now)
+        return driving_power
+
+    def get_next_update_time(self, now: int) -> int:
+        _, next_update_time = self.get_state(now)
+        return next_update_time
+
+    def update(self, info: InfoForDevice) -> DeviceResponse:
```

Comment:
Why do you need this function here? I would presume that the one in ElectricVehicle is sufficient

Reply:
Yes, you're probably right. I've removed it, because I don't see the reason for it also.